### PR TITLE
Enable basic sysctl tuning of HyperShift hosted cluster nodes

### PR DESCRIPTION
**Pull request by dagrayvid**
/cc @jmencak 

This PR enables the Node Tuning operator to run in HyperShift environments, to manage the tuning of hosted cluster nodes.

**A brief explanation of the design:**
The operator will run in the management cluster, in the hosted cluster control plane namespace, alongside some other cluster operators like OLM, network operator, etc. It will manage Tuned / Profile objects and a Tuned daemonset that exist in the hosted cluster API, but are subordinate to Tuned YAMLs in ConfigMaps in the management cluster hosted control plane namespace. To create Tuned in this environment, a user will create a ConfigMap which contains a Tuned manifest in the data.tuned field, and then reference this ConfigMap in the NodePool API spec.tunedConfig field. The NTO operator watches for the ConfigMap in the hosted cluster control plane namespace, and reconciles Tuned objects in the hosted cluster accordingly. See the corresponding HyperShift PR: https://github.com/openshift/hypershift/pull/1651

This PR is "phase-1" of NTO enablement on HyperShift and only supports basic sysctl settings. It does not support kernel args (depends on MachineConfig API), and it does not support MachineConfigPool matching. We plan to enable kernel parameter functionality in "phase-2" which will require a follow-up PR to NTO and the HyperShift repo.

**How to try it out:**
First, build the HyperShift binary and install HyperShift similarly to how it is explained in the [HyperShift docs](https://hypershift-docs.netlify.app/getting-started/), but with a custom image.

```
git clone https://github.com/dagrayvid/hypershift.git
cd hypershift
git checkout nto-poc
make build
sudo install -m 0755 bin/hypershift /usr/local/bin/hypershift

hypershift install --hypershift-image=quay.io/dagray/hypershift-operator:nto-poc --oidc-storage-provider-s3-bucket-name $BUCKET_NAME --oidc-storage-provider-s3-credentials $AWS_CREDS --oidc-storage-provider-s3-region $REGION
```

Then create a HyperShift cluster, using the same custom image as the control-plane-operator image:
```
hypershift create cluster aws --name $CLUSTER_NAME --node-pool-replicas=2 --base-domain $BASE_DOMAIN --pull-secret $PULL_SECRET --aws-creds $AWS_CREDS --region $REGION --generate-ssh --control-plane-operator-image quay.io/dagray/hypershift-operator:nto-poc
```

This should automatically deploy NTO in the control-plane namespace along with the other resources. 

An example of how to build a custom HyperShift image from my fork:
```
git clone https://github.com/dagrayvid/hypershift.git
cd hypershift
git checkout nto-poc
RUNTIME=podman AUTHFILE="--authfile=<PATH_TO_AUTHFILE>" IMG=quay.io/dagray/hypershift-operator:nto-poc make docker-build
podman push --authfile=<PATH_TO_AUTHFILE> quay.io/dagray/hypershift-operator:nto-poc
```

As an example of how to apply some custom tuning to the hosted cluster nodes, you can create the following ConfigMap which contains a Tuned object inside of it:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: hypershift-tuned-test
  namespace: clusters
data:
  tuned: |
    apiVersion: tuned.openshift.io/v1
    kind: Tuned
    metadata:
      name: openshift-dummy
      namespace: openshift-cluster-node-tuning-operator
    spec:
      profile:
      - data: |
          [main]
          summary=Custom OpenShift profile
          include=openshift-node

          [sysctl]   # Section with minimum changes
          net.ipv4.ip_local_port_range="1024 65532"
        name: openshift-dummy
      recommend:
      - match:
        - label: dummy-profile
        priority: 20
        profile: openshift-dummy
```
Note the metadata.namespace field. This should be the same namespace as the NodePool object that it will correspond to.

Reference this ConfigMap in the spec.tunedConfig field of the NodePool:
`oc patch -n clusters nodepool dagray-hypershift-1-us-east-1a --type merge -p '{"spec":{"tunedConfig":[{"name": "hypershift-tuned-test"}]}}'`

From there, the nodepool controller should reconcile this, embedding the tuned configuration into the tuned ConfigMap  in the control plane namespace. For example:

```
$ oc get configmaps -n clusters-dagray-hypershift-1 | grep tuned
tuned-dagray-hypershift-2-us-east-1a   1      22m
```
The data.tuned field of this ConfigMap will contain the Tuned manifests referenced in the NodePool.

Then, in our hosted cluster if we label a node with dummy-profile="", we should see this tuning applied to the node.

### CODE REVIEW

**Comment by jmencak** [comment]
For those trying this PR, make sure you use [David's hypershift](https://github.com/dagrayvid/hypershift.git) for installation.

**Comment by dagrayvid** [comment]
> For those trying this PR, make sure you use [David's hypershift](https://github.com/dagrayvid/hypershift.git) for installation.

Thanks Jiri, I updated the PR description to note that the binary should be built before installing hypershift into the management cluster.

**Comment by dagrayvid** [comment]
@jmencak @liqcui,  I added some steps to the PR description to explain how to apply custom (sysctl) Tuning to a node in the hosted cluster once you get NTO installed.

**Comment by jmencak** [comment]
@dagrayvid first of all, thank you for the PR.  I've looked at your code and verified it works.  I believe this is a good PoC.  However, prior to going further with this, we should consider the following:
  * getting some feedback on your `hypershift` [installer mods](https://github.com/dagrayvid/hypershift/tree/nto-poc) from HyperShift maintainers
  * see if we could leaverage (OpenShift CoreOS layering)[https://github.com/openshift/enhancements/pull/1032] to run `TuneD` in the HyperShift environment

@MarSik any other thoughts?

**Review by jmencak — COMMENTED** [review]
The proposed changes look good to me, only a few last minor nits.

**Review by jmencak — COMMENTED** [review]
Thank you for the changes, David.  They make sense to me and I've also tested they work.  One additional request to the other comments.  Where possible can we try to avoid using *not* in if/then/else clauses?  Example:
```
if !ntoconfig.InHyperShift() {
  // Not in HyperShift
} else {
  // In HyperShift
}
```
vs. IMO easier to read:
```
if ntoconfig.InHyperShift() {
  // In HyperShift
} else {
  // Not in HyperShift
}
```

File: pkg/operator/controller.go [hunk 1]
```diff
@@ -145,6 +152,18 @@ func NewController() (*Controller, error) {
...
+	if ntoconfig.InHyperShift() {
+		managementKubeconfig, err := ntoclient.GetInClusterConfig()
+		if err != nil {
+			return nil, err
+		}
+		controller.clients.ManagementKube, err = kubeset.NewForConfig(managementKubeconfig)
+		if err != nil {
+			return nil, err
+		}
+		controller.clients.ManagementDynamic, err = dynamic.NewForConfig(managementKubeconfig)
```
[the remark this answers is not in the document]
Comment (dagrayvid) [reply pkg/operator/controller.go:164]: Oh good catch, sorry about that... Fixed now.

File: pkg/operator/hypershift.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/klog/v2"
+
+	tunedv1 "github.com/openshift/cluster-node-tuning-operator/pkg/apis/tuned/v1"
+	ntoconfig "github.com/openshift/cluster-node-tuning-operator/pkg/config"
+)
+
+// TODO remove unnecessary debugging log lines.
```
[the remark this answers is not in the document]
Comment (dagrayvid) [reply pkg/operator/hypershift.go]: Sorry, I forgot to address this in my updates yesterday! Fixed now.

File: cmd/cluster-node-tuning-operator/main.go [hunk 3]
```diff
@@ -100,16 +102,18 @@ func operatorRun() {
...
+	// 2. None namespace: for cluster-wide resources
+	ntoNamespace := config.WatchNamespace()
 	namespaces := []string{
 		ntoNamespace,
 		metav1.NamespaceNone,
 	}
 
 	restConfig := ctrl.GetConfigOrDie()
 	le := util.GetLeaderElectionConfig(restConfig, enableLeaderElection)
+
```
[the remark this answers is not in the document]
Comment (dagrayvid) [reply cmd/cluster-node-tuning-operator/main.go]: Done.

File: pkg/operator/hypershift.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	hcTuneds := hcTunedList.Items
+	hcTunedMap := tunedMapFromList(hcTuneds)
+	cmTunedMap := tunedMapFromList(cmTuneds)
+
+	for tunedName, cmTuned := range cmTunedMap {
+		if hcTuned, ok := hcTunedMap[tunedName]; ok {
+			klog.V(1).Infof("hosted cluster already contains Tuned %v from ConfigMap", tunedName)
+			if reflect.DeepEqual(cmTuned.Spec.Profile, hcTuned.Spec.Profile) &&
+				reflect.DeepEqual(cmTuned.Spec.Recommend, hcTuned.Spec.Recommend) {
+				klog.V(2).Infof("Hosted cluster version of Tuned %v matches the ConfigMap %s config", tunedName, cmTuned.ObjectMeta.Name)
```
[the remark this answers is not in the document]
Comment (dagrayvid) [reply pkg/operator/hypershift.go]: I think I fixed all the occurrences of this issue in this PR.

File: pkg/operator/profilecalculator.go [hunk 5]
```diff
@@ -208,6 +212,70 @@ func (pc *ProfileCalculator) calculateProfile(nodeName string) (string, map[stri
...
+		return "", "", operand, fmt.Errorf("failed to list Tuneds in NodePool %s: %v", nodePoolName, err)
+	}
+	defaultTuned, err := pc.listers.TunedResources.Get(tunedv1.TunedDefaultResourceName)
+	if err != nil {
+		return defaultProfile, "", operand, fmt.Errorf("failed to get Tuned %s: %v", tunedv1.TunedDefaultResourceName, err)
+	}
+	tunedList = append(tunedList, defaultTuned)
+
+	for _, recommend := range tunedRecommend(tunedList) {
+		// Start with node/pod label based matching to MachineConfig matching when
```
[the remark this answers is not in the document]
Comment (dagrayvid) [reply pkg/operator/profilecalculator.go]: Fixed now.

**Comment by jmencak** [comment]
Thank you David, the changes
/lgtm

**Pull request merged**